Wrap Op params for many gpuarray DNN Ops and add cuDNN v6 integration.

notoraptor opened the pull request:
This PR adds and extends a lot of work done in #5866 . ParamsType is now used for many gpuarray DNN ops, and cuDNN v6 integration is extended.

Ops currently rewritten in this PR:

 - GpuDnnConvDesc
 - GpuDnnConv
 - GpuDnnConvGradW
 - GpuDnnConvGradI
 - GpuDnnPool
 - GpuDnnPoolGrad
 - GpuDnnBatchNormInference
 - GpuDnnBatchNormGrad

cuDNN v6 integration:

 - Support MAX DETERMINISTIC algorithm for GpuDnnPool with cuDNN v6.
 - Update pooling tests for DNN module so that they use the available algorithms depending on runtime cuDNN version.
 - Allow CPU Pool and PoolGrad ops to use MAX_DETERMINISTIC algo when cuDNN v6 is used with GPU counterparts.
 - Encapsulate cuDNN constants used in DNN module, to help choose the right cuDNN definitions depending on the runtime cuDNN version. Currently supported cuDNN versions: v5.1, v6.0. Encapsulation is done in module `cudnn_defs.py`.

Next steps:
 - Add a new tests file to do exhaustive tests on gpuarray DNN ops. I had created a separate PR for that, but maybe I will have to either continue this work here (in this PR), or wait for this PR to be merged first.
 - Check if I can/must add other specific checkings that depend on cuDNN version (e.g. check if an algorithm chosen at runtime can be really applied on current inputs).
 - Check if I can also rewrite GpuDnnBatchnorm (maybe in another PR). This op is a little hard to rewrite because I can't convert some macros (used in C code) to params.

But at this point, this PR can be reviewed, @lamblin @nouiz @abergeron .

notoraptor commented:
Don't forget this PR, @lamblin !

abergeron commented:
I think I'm supposed to review this, but I've been putting it off due to a fever which makes it hard to concentrate.  I'm getting better though so maybe tonight or next Monday.

notoraptor commented:
Rebased to integrate last changes in corr(3d)MM ops + one new commit for a small fix.

abergeron commented on theano/gpuarray/dnn.py:
```diff
@@ -129,13 +136,12 @@ def _dnn_check_compile():
                       "Theano. If you encounter problems, try "
                       "updating Theano or downgrading cuDNN to "
-                      "version 6.0.")
+                      "a version >= v5 and < v6.1.")
```
I would use <= v6 as the upper limit since we don't know if there will be a v6.1 yet.

notoraptor commented:
Updated. I have also added a TODO mention in some comments to remind that we need to  test new added convolution algorithms (mainly "winograd non fused").

I have also removed the comment that said new pooling algorithm `max_deterministic` is not documented: in fact, it is documented in dnn_pool() docstring: https://github.com/Theano/Theano/pull/5964/files#diff-e0ca808f4cca44e42378ba7aced1ab8cR1429 .
